src/realm/object-store/class.hpp

```diff
@@ -0,0 +1,56 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2023 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+#ifndef REALM_OS_CLASS_HPP
+#define REALM_OS_CLASS_HPP
+
+#include <realm/obj.hpp>
+#include "realm/parser/keypath_mapping.hpp"
+#include "realm/parser/query_parser.hpp"
+#include <memory>
+
+namespace realm {
+
+class ObjectSchema;
+class Realm;
+class Query;
+
+class Class {
+public:
+    Class(const std::shared_ptr<Realm>& r, StringData object_type);
+
+    size_t size() const;
+    TableKey get_key() const;
+    ColKey get_column_key(StringData) const;
+    Query get_query(const std::string& query_string, query_parser::Arguments& args,
+                    const query_parser::KeyPathMapping& mapping) const;
+
+    Obj create_object(Mixed primary_value = {});
+    Obj get_object(Mixed primary_value);
```

Comment:
We also need a `get_object(ObjKey)`  and `try_get_object(ObjKey)` (or omit the non-try version).

You can see what the JS SDK is currently using on (Const)TableRef in this image:
<img width="564" alt="image" src="https://user-images.githubusercontent.com/118456/233596941-4ae6faa6-ab09-4742-880a-cbddf8b411cb.png">

While we don't currently use the iterator, we probably should. I suspect it is faster than iterating the `Results` especially since I think we do that by indexing with an incrementing index.

Reply:
Additional uses of `TableRef` include: 

* `Helpers::get_or_create_object_with_primary_key()`  as mentioned in [another comment](https://github.com/realm/realm-core/pull/6521#discussion_r1173547349)
* `Obj::get_backlink_view(TableRef, CollKey) -> TableView` which is then passed to the `Results` constructor. This is also the only use of `TableView`, so a more direct way to work with backlinks would remove that too.
* `obj.get_table().get_key()` If we had a direct way to get the `TableKey` from an `Obj` we wouldn't need this.
* `results.get_query().get_table().query()` We need a way to add additional filtering to a `Results` using a query string, ideally without building a `Query` object that we only use to hand back to core.
** Ideally we could also have a method like `Class::queryResults(queryString) -> Results` and `Results::description() -> string`, then we could remove the `Query` type from the binding completely (not sure if other sdks use it though)